### ISSUE TITLE
Build error

### ISSUE THREAD

**Issue by tdegeus** [xtensor-stack/xsimd#335]
On my CI I'm getting

```
In file included from /usr/lib/gcc/x86_64-linux-gnu/5/include/immintrin.h:45:0,
1086                 from /usr/lib/gcc/x86_64-linux-gnu/5/include/x86intrin.h:46,
1087                 from /home/travis/miniconda/include/xsimd/config/xsimd_include.hpp:19,
1088                 from /home/travis/miniconda/include/xsimd/types/xsimd_types_include.hpp:14,
1089                 from /home/travis/miniconda/include/xsimd/types/xsimd_traits.hpp:17,
1090                 from /home/travis/miniconda/include/xsimd/xsimd.hpp:16,
1091                 from /home/travis/miniconda/include/xtensor/xtensor_config.hpp:84,
1092                 from /home/travis/miniconda/include/xtensor/xbuffer_adaptor.hpp:21,
1093                 from /home/travis/miniconda/include/xtensor/xtensor.hpp:19,
1094                 from /home/travis/build/tdegeus/GooseEYE/include/GooseEYE/config.h:26,
1095                 from /home/travis/build/tdegeus/GooseEYE/include/GooseEYE/GooseEYE.h:10,
1096                 from /home/travis/build/tdegeus/GooseEYE/test/main.cpp:2:
1097/home/travis/miniconda/include/xsimd/types/xsimd_avx512_int64.hpp: In function ‘xsimd::batch<long int, 8ul> xsimd::operator<<(const xsimd::batch<long int, 8ul>&, int32_t)’:
1098/home/travis/miniconda/include/xsimd/types/xsimd_avx512_int64.hpp:412:16: error: the last argument must be an 8-bit immediate
1099         return _mm512_slli_epi64(lhs, rhs);
1100                ^
1101/home/travis/miniconda/include/xsimd/types/xsimd_avx512_int64.hpp: In function ‘xsimd::batch<long int, 8ul> xsimd::operator>>(const xsimd::batch<long int, 8ul>&, int32_t)’:
1102/home/travis/miniconda/include/xsimd/types/xsimd_avx512_int64.hpp:417:16: error: the last argument must be an 8-bit immediate
1103         return _mm512_srli_epi64(lhs, rhs);
1104                ^
```

FYI: https://travis-ci.org/tdegeus/GooseEYE/jobs/623781528

**Comment by JohanMabille** [comment]
Does it constantly fail or did you manage to have it working at some point? Also does it fail only for this compiler?

**Comment by JohanMabille** [comment]
Apparently, the shift intrinsics require the shift parameter to be `immediate`, i.e., known at compile time. Since the build succeeds for other version of GCC? it might be that GCC 5 does not manage to inline the wrapping function and therefore does not see the parameter as an immediate.

We might solve this by forcing the inlining.

**Comment by tdegeus** [comment]
I only had it on one GCC version and not for all hardware configurations: the same GCC also ran successfully on another build.

**Comment by JohanMabille** [comment]
Can you add the following to your CI script before you start the build?

```bash
    - if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then
        cat /proc/cpuinfo;
        ldd --version;
      fi
```

I am pretty sure that the failing builds will be on Intel `Xeon model 85`, while builds on `Xeon model 63` will always be fine.

We have a similar issue with the build of `xtensor`, although the symptoms are a bit different, but I suspect a common root cause.

**Comment by tdegeus** [comment]
Will do. In the mean time I am also experiencing this problem on GCC6

**Comment by tdegeus** [comment]
```
if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then cat /proc/cpuinfo; ldd --version; fi
processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 85
model name	: Intel(R) Xeon(R) CPU
stepping	: 7
microcode	: 0x1
cpu MHz		: 2800.202
cache size	: 33792 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc eagerfpu pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f rdseed adx smap clflushopt clwb avx512cd xsaveopt xsavec xgetbv1 arat
bugs		:
bogomips	: 5600.40
clflush size	: 64
cache_alignment	: 64
address sizes	: 46 bits physical, 48 bits virtual
power management:

processor	: 1
vendor_id	: GenuineIntel
cpu family	: 6
model		: 85
model name	: Intel(R) Xeon(R) CPU
stepping	: 7
microcode	: 0x1
cpu MHz		: 2800.202
cache size	: 33792 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 1
initial apicid	: 1
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc eagerfpu pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f rdseed adx smap clflushopt clwb avx512cd xsaveopt xsavec xgetbv1 arat
bugs		:
bogomips	: 5600.40
clflush size	: 64
cache_alignment	: 64
address sizes	: 46 bits physical, 48 bits virtual
power management:

ldd (Ubuntu EGLIBC 2.19-0ubuntu6.13) 2.19
Copyright (C) 2014 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.
```

https://travis-ci.com/tdegeus/GooseEYE/jobs/266444789

**Comment by JohanMabille** [comment]
I finally got it. There are several factors leading to this issue, and to the random failures we experienced on the other repos of the stack. The common factor is that travisci has made available new machines that support AVX512 instructions set (but did not remove the older ones). If such a machine is selected, the build can fail for many reasons:
- Incompatibility of old binutils. The default configuration of travisci uses the Trusty distribution, which cames with binutils 2.19. This version is not compatible with this instruction set and leads to failure when building gtest.
- several bugs in AVX512 implementation that were not caught by the CI. These are being fixed in #338 
- AVX512 instruction set is supported starting with GCC 6, which was not taking into account in `xsimd_instruction_set.hpp`. This will also be fixed in #338.

I will release xsimd and upgrade the stack after that so that downstream projects do not have these issues anymore.
